lib/internal/fs/watch/linux.js

```diff
@@ -0,0 +1,244 @@
+
+    try {
+      const files = await opendir(folder);
+
+      for await (const file of files) {
+        const f = path.join(folder, file.name);
+
+        if (this.#closed) {
+          break;
+        }
+
+        if (!this.#files.has(f)) {
+          this.#files.set(f, file);
+          this.emit('change', 'rename', this.#getPath(f));
+
+          if (file.isDirectory()) {
+            await this.#watchFolder(f);
+          } else {
+            this.#watchFile(f);
+          }
+        }
+      }
+    } catch (error) {
+      this.emit('error', error);
+    }
+  }
+
+  #watchFile(file) {
+    validateString(file, 'file');
+
+    const { watchFile } = lazyLoadFsSync();
+
+    if (this.#closed) {
+      return;
+    }
+
+    const existingStat = this.#files.get(file);
+
+    watchFile(file, {
+      persistent: this.#options.persistent,
+    }, (statWatcher, previousStatWatcher) => {
+      if (existingStat && !existingStat.isDirectory() &&
+        statWatcher.nlink !== 0 && existingStat.mtime?.getTime() === statWatcher.mtime?.getTime()) {
+        return;
+      }
+
+      this.#files.set(file, statWatcher);
+
+      if (statWatcher.birthtimeMs === 0 && previousStatWatcher.birthtimeMs !== 0) {
+        // The file is now deleted
+        this.#files.delete(file);
+        this.emit('change', 'rename', this.#getPath(file));
+
+        if (statWatcher.isDirectory()) {
+          this.#unwatchFolder(file);
+        }
+      } else if (statWatcher.isDirectory()) {
+        this.emit('change', 'change', this.#getPath(file));
+        this.#watchFolder(file);
+      } else {
+        this.emit('change', 'change', this.#getPath(file));
+      }
+    });
+  }
+
+  [kFSWatchStart](filename) {
+    validateString(filename, 'filename');
+
+    if (filename.length === 0) {
+      throw new ERR_INVALID_ARG_TYPE('filename');
+    }
+
+    const file = lazyLoadFsSync().statSync(filename);
+
+    this.#rootPath = filename;
+    this.#closed = false;
+    this.#files.set(filename, file);
+
+    if (file.isDirectory()) {
+      traverse(filename, this.#files)
+        .then(() => {
+          for (const f of this.#files.keys()) {
+            this.#watchFile(f);
+          }
+        });
+    } else {
+      this.#watchFile(filename);
+    }
+  }
+
+  ref() {
+    // This is kept to have the same API with FSWatcher
+    throw new ERR_FEATURE_UNAVAILABLE_ON_PLATFORM('ref');
+  }
+
+  unref() {
+    // This is kept to have the same API with FSWatcher
+    throw new ERR_FEATURE_UNAVAILABLE_ON_PLATFORM('unref');
+  }
+
+  async next() {
+    if (this.#closed) {
+      return { __proto__: null, done: true };
+    }
+
+    const result = await new Promise((resolve) => {
+      this.once('change', (eventType, filename) => {
+        resolve({ eventType, filename });
+      });
+    });
+
+    return { __proto__: null, done: false, value: result };
+  }
+
+  // eslint-disable-next-line require-yield
+  *[SymbolAsyncIterator]() {
+    return this;
```

Comment:
I think you'll get better results with the following implementation:
```suggestion
  [SymbolAsyncIterator]() {
    const watcher = this;
    return ObjectCreate({
      __proto__: null,
      async next() {
        if (watcher.#closed) {
          return { __proto__: null, done: true };
        }
        const { 0: eventType, 1: filename } = await once(watcher, 'change');
        return { __proto__: null, value: { __proto__: null, eventType, filename } };
      },
      [Symbol.asyncIterator]() { return this; },
    });
  }
```

Reply:
Thank you. I'll update the pull request.

Reply:
This changes are making test-fs-watch-recursive-linux-promise.js test fail. This was the main reason I added a `*` to SymbolAsyncIterator, because it was keeping the event-loop open (I guess). Do you have any idea why this is happening @aduh95? 

```
Mismatched noop function calls. Expected exactly 1, actual 0.
    at Proxy.mustCall (/home/parallels/Desktop/node/test/common/index.js:406:10)
    at Object.<anonymous> (/home/parallels/Desktop/node/test/parallel/test-fs-watch-recursive-linux-promise.js:43:18)
    at Module._compile (node:internal/modules/cjs/loader:1159:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1213:10)
    at Module.load (node:internal/modules/cjs/loader:1037:32)
    at Module._load (node:internal/modules/cjs/loader:878:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:82:12)
    at node:internal/main/run_main_module:23:47
Command: out/Release/node /home/parallels/Desktop/node/test/parallel/test-fs-watch-recursive-linux-promise.js
```